### ISSUE TITLE
[typescript-eslint/no-unsafe-call] false positive in vs code when editing

### ISSUE THREAD

**Issue by krisztianb** [typescript-eslint/typescript-eslint#3012]
- [ ] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

This issue is a revisiting of the one created by @sfabriece here: #2885 

I managed to find a way to reproduce this all of the time. It's pretty easy. I'm not sure if this is a bug in this package or the VSCode GUI extension though, because it involves certain specific editing steps to perform.

1. Create a file that exports anything (eg a function).
2. Create another file that imprts that function and calls it.
3. Save all files.
3. Go back to the file with the function definition.
4. Press F2 and rename the function.
5. Voila: The error "Unsafe call of an any typed value.eslint@typescript-eslint/no-unsafe-call" is shown in the file that calls the function.

**Expected Result**

After Step 5 there should be no error displayed in the files that are importing the renamed function.

**Actual Result**

After Step 5 all files importing the function show linting error messages.

**Additional Info**

The error is only shown in the GUI and is gone when VSCode is restarted.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.2` |
| `@typescript-eslint/parser`        | `4.14.2` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.19.0` |
| `node`                             | `12.13.1` |

**Comment by bradzacher** [comment]
Without doing any testing - from your description it's pretty obvious that this is a race condition.

There's a race between the order in which the files are saved and the order in which vscode asks us to lint files - i.e. outside our control.

In which case there's likely not a lot that we can do about it.

---

If you set `parserOptions.debugLevel = true`, then you should be able to grab the logs from the vscode eslint extension logs and dump them here.

**Comment by krisztianb** [comment]
Well the strange thing is, the order in which the files are saved does not matter. I must have mead a mistake there, because if I save the files one by one now, the error is still displayed.

This is the eslint log output:

```
2021-02-11T08:22:23.956Z eslint:cli-engine Lint E:\code\web\master\calclib\open_document.ts
2021-02-11T08:22:23.956Z eslint:linter Linting code for E:\code\web\master\calclib\open_document.ts (pass 1)
2021-02-11T08:22:23.956Z eslint:linter Verify
2021-02-11T08:22:23.956Z eslint:linter With ConfigArray: E:\code\web\master\calclib\open_document.ts
2021-02-11T08:22:23.957Z typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set { 'e:\\code\\web\\master\\tsconfig.json' }
2021-02-11T08:22:23.957Z typescript-eslint:typescript-estree:createProjectProgram Creating project program for: E:\code\web\master\calclib\open_document.ts
2021-02-11T08:22:23.957Z typescript-eslint:typescript-estree:createWatchProgram Found existing program for file. e:\code\web\master\calclib\open_document.ts
2021-02-11T08:22:24.004Z typescript-eslint:parser:parser Resolved libs from program: [ 'es2018.full' ]
2021-02-11T08:22:24.079Z typescript-eslint:eslint-plugin:utils:types Found an "error" any type
2021-02-11T08:22:24.204Z eslint:linter Generating fixed text for E:\code\web\master\calclib\open_document.ts (pass 1)
2021-02-11T08:22:24.204Z eslint:source-code-fixer Applying fixes
2021-02-11T08:22:24.204Z eslint:source-code-fixer shouldFix parameter was false, not attempting fixes
2021-02-11T08:22:24.204Z eslint:cli-engine Linting complete in: 251ms
2021-02-11T08:22:24.211Z eslint:cli-engine Lint E:\code\web\master\calclib\recalcables\history_document.ts
2021-02-11T08:22:24.211Z eslint:linter Linting code for E:\code\web\master\calclib\recalcables\history_document.ts (pass 1)
2021-02-11T08:22:24.211Z eslint:linter Verify
2021-02-11T08:22:24.211Z eslint:linter With ConfigArray: E:\code\web\master\calclib\recalcables\history_document.ts
2021-02-11T08:22:24.212Z typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set { 'e:\\code\\web\\master\\tsconfig.json' }
2021-02-11T08:22:24.212Z typescript-eslint:typescript-estree:createProjectProgram Creating project program for: E:\code\web\master\calclib\recalcables\history_document.ts
2021-02-11T08:22:24.212Z typescript-eslint:typescript-estree:createWatchProgram Found existing program for file. e:\code\web\master\calclib\recalcables\history_document.ts
2021-02-11T08:22:24.239Z typescript-eslint:parser:parser Resolved libs from program: [ 'es2018.full' ]
2021-02-11T08:22:24.336Z eslint:linter Generating fixed text for E:\code\web\master\calclib\recalcables\history_document.ts (pass 1)
2021-02-11T08:22:24.336Z eslint:source-code-fixer Applying fixes
2021-02-11T08:22:24.336Z eslint:source-code-fixer shouldFix parameter was false, not attempting fixes
2021-02-11T08:22:24.336Z eslint:cli-engine Linting complete in: 129ms
```

Does this help?

**Comment by bradzacher** [comment]
The logs look pretty normal for the most part - you can see that for both files we successfully find the file in one of our existing programs.

However when linting `E:\code\web\master\calclib\open_document.ts`, you can see the following log
```
typescript-eslint:eslint-plugin:utils:types Found an "error" any type
```

This log line indicates that when checking to see if a type was `any`, we found that TS had created an `any` and marked it as an "error".
TS does this whenever it encounters a type that it cannot resolve.

There are many complex interactions between multiple tools (vscode, eslint, our tooling, and typescript), and more complexity with dealing with disk state vs editor state.
There's a lot going on here, and TBH I don't think there's anything we can do about it right now.

If you save the exporting file, and then save (and maybe close and reopen) the importing file - then it should all sync up fine.
If you're ever in doubt - save the files and reload vscode (ctrl + shift + p > reload), or save the files and run the CLI on the file `eslint <path to file>`.

**Comment by krisztianb** [comment]
Thanks for the feedback.

Reopening the file that does the import after saving it makes the error not show up again.

I guess I only reported this, because this seems to be the only case/rule in which I've experienced such a behavior so far and I was wondering whether this is normal or not.

Closing this a something that "can't" be fixed and is not a bug.

**Comment by sfabriece** [comment]
@krisztianb it might be worth it to figure out which tool first reports the error and opening up the issue to the respective maintainers. For example the typescript team. I would imagine that this would be something they would be interested in fixing.

**Comment by krisztianb** [comment]
@sfabriece sounds interesting. How would you go about doing that?

**Comment by bradzacher** [comment]
> I guess I only reported this, because this seems to be the only case/rule in which I've experienced such a behavior so far and I was wondering whether this is normal or not.

The `no-unsafe-*` class of rules will highlight this problem more than others because they report when you operate on an `any`, so they are particularly sensitive to when an "error" any sneaks in.

Moat other rules will just ignore `any`s.

**Comment by sfabriece** [comment]
@krisztianb I would create a minimum repro and only enable the tools and extensions to reproduce the errors. Then enable logging in the tools to figure out how the errors manifest themselves.

Ideally, that should be enough to identify the culprit 😄

**Comment by krisztianb** [comment]
@sfabriece I see. To be honest currently I don't feel the motivation to do this. Might take a shot when this issue becomes more annoying.

**Comment by sfabriece** [comment]
@krisztianb same here. I'll be sure to ping you if I get started on it.
